File: VB/Dashboard_CustomExtractDriver/ExtractEncryptionDriver.vb
Imports System
Imports System.Collections.Generic
Imports System.Security.Cryptography
Imports DevExpress.DashboardCommon

Namespace Dashboard_CustomExtractDriver
    Public Class ExtractEncryptionDriver
        Implements ICustomExtractDriver
        Public Function CreateReadSession(ByVal resourceName As String) As  _
            IDriverReadSession Implements ICustomExtractDriver.CreateReadSession
            Return New EncryptionReadSession(FileExtractDriver.Instance.CreateReadSession(resourceName))
        End Function
        Public Function CreateWriteSession(ByVal resourceName As String) As  _
            IDriverWriteSession Implements ICustomExtractDriver.CreateWriteSession
            Return New EncryptionWriteSession(FileExtractDriver.Instance.CreateWriteSession(resourceName))
        End Function
    End Class

    Public Class EncryptionWriteSession
        Implements IDriverWriteSession
        Private writeSession As IDriverWriteSession
        Public Sub New(ByVal writeSession As IDriverWriteSession)
            Me.writeSession = writeSession
        End Sub
        Public Sub Dispose() Implements System.IDisposable.Dispose
            writeSession.Dispose()
        End Sub
        Public Sub SetPage(ByVal pageID As Guid, ByVal data() As Byte) Implements IDriverWriteSession.SetPage
            writeSession.SetPage(pageID, Encrypt(data))
        End Sub
        Private Function Encrypt(ByVal page() As Byte) As Byte()
            Dim entropy() As Byte = {1, 2, 3}
            Dim encryptedPage() As Byte = ProtectedData.Protect(page, entropy, DataProtectionScope.CurrentUser)
            Return encryptedPage
        End Function
    End Class

    Public Class EncryptionReadSession
        Implements IDriverReadSession
        Private readSession As IDriverReadSession
        Public Sub New(ByVal readSession As IDriverReadSession)
            Me.readSession = readSession
        End Sub
        Public Sub Dispose() Implements System.IDisposable.Dispose
            readSession.Dispose()
        End Sub
        Private Function Decrypt(ByVal page() As Byte) As Byte()
            Dim entropy() As Byte = {1, 2, 3}
            Dim decryptedPage() As Byte = ProtectedData.Unprotect(page, entropy, DataProtectionScope.CurrentUser)
            Return decryptedPage
        End Function
        Public Function GetPage(ByVal pageID As Guid) As Byte() Implements IDriverReadSession.GetPage
            Return Decrypt(readSession.GetPage(pageID))
        End Function
        Public Function GetPageIDs() As IEnumerable(Of Guid) Implements IDriverReadSession.GetPageIDs
            Return readSession.GetPageIDs()
        End Function
    End Class
End Namespace

File: CS/Dashboard_CustomExtractDriver/ExtractEncryptionDriver.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using DevExpress.DashboardCommon;

namespace Dashboard_CustomExtractDriver {
    public class ExtractEncryptionDriver : ICustomExtractDriver {

        public IDriverReadSession CreateReadSession(string resourceName) {
            return new EncryptionReadSession(FileExtractDriver.Instance.CreateReadSession(resourceName));
        }
        public IDriverWriteSession CreateWriteSession(string resourceName) {
            return new EncryptionWriteSession(FileExtractDriver.Instance.CreateWriteSession(resourceName));
        }
    }

    public class EncryptionWriteSession : IDriverWriteSession {
        IDriverWriteSession writeSession;

        public EncryptionWriteSession(IDriverWriteSession writeSession) {
            this.writeSession = writeSession; 
        }
        public void Dispose() {
            writeSession.Dispose();
        }
        public void SetPage(Guid pageID, byte[] data) {
            writeSession.SetPage(pageID, Encrypt(data));
        }
        byte[] Encrypt(byte[] page) {
            byte[] entropy = { 1, 2, 3 };
            byte[] encryptedPage = ProtectedData.Protect(page, entropy, DataProtectionScope.CurrentUser);
            return encryptedPage;
        }
    }

    public class EncryptionReadSession : IDriverReadSession {
        IDriverReadSession readSession;
        public EncryptionReadSession(IDriverReadSession readSession) {
            this.readSession = readSession;
        }
        public void Dispose() {
            readSession.Dispose();
        }
        byte[] Decrypt(byte[] page) {
            byte[] entropy = { 1, 2, 3 };
            byte[] decryptedPage = ProtectedData.Unprotect(page, entropy, DataProtectionScope.CurrentUser);
            return decryptedPage;
        }
        public byte[] GetPage(Guid pageID) {
            return Decrypt(readSession.GetPage(pageID));
        }
        public IEnumerable<Guid> GetPageIDs() {
            return readSession.GetPageIDs();
        }
    }
}